place_tracker/lib/place_map.dart

```diff
@@ -0,0 +1,423 @@
+
+  Future<void> _pushPlaceDetailsScreen(Marker marker) async {
+    assert(marker != null);
+
+    Navigator.push(
+      context,
+      MaterialPageRoute(builder: (context) {
+        return PlaceDetails(
+          place: _places[marker],
+          onChanged: (Place value) {
+            _updatePlaceAndMarker(value, marker);
+          },
+        );
+      }),
+    );
+  }
+
+  Future<void> _updatePlaceAndMarker(Place place, Marker marker) async {
+    _places[marker] = place;
+
+    // Set marker visibility to false to ensure the info window is hidden. Once
+    // the plugin fully supports the Google Maps API, use hideInfoWindow()
+    // instead.
+    await mapController.updateMarker(
+      marker,
+      MarkerOptions(
+        visible: false,
+      ),
+    );
+    await mapController.updateMarker(
+      marker,
+      MarkerOptions(
+        infoWindowText: InfoWindowText(
+          place.name,
+          place.starRating != 0
+            ? '${place.starRating.toString()} Star Rating'
+            : null,
+        ),
+        visible: true,
+      ),
+    );
+  }
+
+  void _updatePlaces(PlaceCategory category) {
+    setState(() {
+      _selectedPlaceCategory = category;
+      _showPlacesForSelectedCategory();
+    });
+  }
+
+  void _showPlacesForSelectedCategory() {
+    _places.forEach((Marker marker, Place place) {
+      mapController.updateMarker(
+        marker,
+        MarkerOptions(
+          visible: place.category == _selectedPlaceCategory,
+        ),
+      );
+    });
+    _zoomToFitPlaces(
+      PlaceUtil.getPlacesForCategory(_selectedPlaceCategory, _places),
+    );
+  }
+
+  void _zoomToFitPlaces(List<Place> places) {
+    // Default min/max values to latitude and longitude of center.
+    double minLat = _center.latitude;
+    double maxLat = _center.latitude;
+    double minLong = _center.longitude;
+    double maxLong = _center.longitude;
+    places.forEach((Place place) {
+      minLat = place.latitude < minLat ? place.latitude : minLat;
+      maxLat = place.latitude > maxLat ? place.latitude : maxLat;
+      minLong = place.longitude < minLong ? place.longitude : minLong;
+      maxLong = place.longitude > maxLong ? place.longitude : maxLong;
+    });
+    mapController.animateCamera(
+      CameraUpdate.newLatLngBounds(
+        LatLngBounds(
+          southwest: LatLng(minLat, minLong),
+          northeast: LatLng(maxLat, maxLong),
+        ),
+        48.0,
+      ),
+    );
+  }
+
+  void _onAddPlaceFabPressed() async {
+    Marker newMarker = await mapController.addMarker(
+      MarkerOptions(
+        position: LatLng(
+          mapController.cameraPosition.target.latitude,
+          mapController.cameraPosition.target.longitude,
+        ),
+        draggable: true,
+        icon: BitmapDescriptor.defaultMarkerWithHue(BitmapDescriptor.hueGreen),
+      ),
+    );
+    setState(() {
+      _pendingMarker = newMarker;
+    });
+  }
+
+  void _confirmAddPlace(BuildContext context) async {
```

Comment:
Not sure why `BuildContext` is a parameter here (you can refer to `State.context` implicitly).  Does this context refer to a different Scaffold?

Reply:
Trying 
`Scaffold.of(State.context).showSnackBar(
  ...
);`
gives me the error
"Instance member 'context' can't be accessed using static access.'

Reply:
By `State.context` I meant the `context` field in `State`.

Reply:
Oh I see. I got this error before I did some refactoring, but I got it to resurface using the context field in State here: 

Scaffold.of() called with a context that does not contain a Scaffold.
E/flutter ( 2186): No Scaffold ancestor could be found starting from the context that was passed to Scaffold.of(). This usually happens when the context provided is from the same StatefulWidget as that whose build function actually creates the Scaffold widget being sought.
E/flutter ( 2186): There are several ways to avoid this problem. The simplest is to use a Builder to get a context that is "under" the Scaffold. For an example of this, please see the documentation for Scaffold.of():
E/flutter ( 2186):   https://docs.flutter.io/flutter/material/Scaffold/of.html
E/flutter ( 2186): A more efficient solution is to split your build function into several widgets. This introduces a new context from which you can obtain the Scaffold. In this solution, you would have an outer widget that creates the Scaffold populated by instances of your new inner widgets, and then in these inner widgets you would use Scaffold.of().
E/flutter ( 2186): A less elegant but more expedient solution is assign a GlobalKey to the Scaffold, then use the key.currentState property to obtain the ScaffoldState rather than using the Scaffold.of() function.

Reply:
```
class _Home extends StatelessWidget {
  const _Home({ Key key }) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return PlaceMap(
      center: const LatLng(45.521563, -122.677433),
    );
  }
}

void main() {
  runApp(
    MaterialApp(
      debugShowCheckedModeBanner: false,
      title: 'Place Tracker',
      home: _Home(),
    );
}
```

Reply:
I added this modification but I am still getting the same error above.

Reply:
Sorry, I didn't really stop to think about that. The change I was suggesting above wasn't needed.

PlaceMap's build function creates its Scaffold, so Scaffold.of() can't find it because: same context.

Using a Builder to construct the Scaffold's body, should solve the problem.

Reply:
I was able to remove the builder from _AddPlaceButtonBar entirely. I still need to pass context into the _confirmAddPlace method because otherwise, it will try to use the State's context, resulting in the before-mentioned error. I am passing the context from _AddPlaceButtonBar's main build method.

Reply:
Ah nevermind I see what you mean after reading your other comment below.